crates/spartan-farmer/src/spartan.rs

```diff
@@ -0,0 +1,168 @@
+#![warn(missing_debug_implementations, missing_docs)]
+//! This is an adaptation of [SLOTH](https://eprint.iacr.org/2015/366) (slow-timed hash function) into a time-asymmetric permutation using a standard CBC block cipher. This code is largely based on the C implementation used in [PySloth](https://github.com/randomchain/pysloth/blob/master/sloth.c) which is the same as used in the paper.
+
+use sloth256_189::cpu;
+use sloth256_189::cuda;
+
+/// Spartan struct used to encode and validate
+#[derive(Debug, Clone)]
+pub struct Spartan {
+    genesis_piece: [u8; 4096],
+}
+
+impl Spartan {
+    /// New instance with 256-bit prime and 4096-byte genesis piece size
+    pub fn new(genesis_piece: [u8; 4096]) -> Self {
+        Spartan { genesis_piece }
+    }
+}
+
+impl Spartan {
+    /// Create an encoding based on genesis piece using provided encoding key hash, nonce and
+    /// desired number of rounds
+    pub fn encode(&self, encoding_key_hash: [u8; 32], nonce: u64, rounds: usize) -> [u8; 4096] {
+        let mut expanded_iv = encoding_key_hash;
+        for (i, &byte) in nonce.to_le_bytes().iter().rev().enumerate() {
+            expanded_iv[32 - i - 1] ^= byte;
+        }
+
+        let mut encoding = self.genesis_piece;
+
+        cpu::encode(&mut encoding, &expanded_iv, rounds).unwrap();
+
+        encoding
+    }
+
+    pub fn batch_encode(
+        &self,
+        piece_amount: usize,
+        pieces: &mut [u8],
+        encoding_key_hash: [u8; 32],
+        nonce_array: &[u64],
+        rounds: usize,
+    ) {
+        // each expanded_iv will be in format [u8; 32], so `piece_amount` expanded_iv's
+        // should consume [u8; 32 * piece_amount] space.
+        let mut expanded_iv_vector: Vec<u8> = Vec::with_capacity(piece_amount * 32);
+        for x in 0..piece_amount {
+            // same encoding_key_hash will be used for each expanded_iv
+            let mut expanded_iv = encoding_key_hash.clone();
```

Comment:
You don't need to clone trivially copyable values:
```suggestion
            let mut expanded_iv = encoding_key_hash;
```

This will result in `clippy` warnings that are treated as errors on `main` branch already.

Run `cargo clippy` to know more.